relay-threading/src/metrics.rs

```diff
@@ -1,18 +1,118 @@
-use relay_statsd::GaugeMetric;
-
-/// Gauge metrics emitted by the asynchronous pool.
-pub enum AsyncPoolGauges {
-    /// Number of futures queued up for execution in the asynchronous pool.
-    AsyncPoolQueueSize,
-    /// Number of futures being driven in each thread of the asynchronous pool.
-    AsyncPoolFuturesPerThread,
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+/// Metrics for a single thread in an asynchronous pool.
+///
+/// This struct provides a way to track and query metrics specific to an individual thread, such as
+/// the number of futures it has polled. It is designed for safe concurrent access across threads.
+#[derive(Debug, Default)]
+pub struct ThreadMetrics {
+    /// Number of futures that are currently being polled concurrently by the thread in the pool.
+    polled_futures: AtomicU64,
+}
+
+impl ThreadMetrics {
+    /// Returns the number of futures polled by the thread.
+    ///
+    /// This method provides a snapshot of the total futures processed by the thread at the time of
+    /// the call. It is useful for monitoring thread activity and workload distribution.
+    pub fn polled_futures(&self) -> u64 {
+        self.polled_futures.load(Ordering::SeqCst)
+    }
+
+    /// Updates the `polled_futures` value.
+    pub(crate) fn update_polled_futures(&self, polled_futures: u64) {
+        self.polled_futures.store(polled_futures, Ordering::SeqCst);
+    }
 }
 
-impl GaugeMetric for AsyncPoolGauges {
-    fn name(&self) -> &'static str {
-        match self {
-            Self::AsyncPoolQueueSize => "async_pool.queue_size",
-            Self::AsyncPoolFuturesPerThread => "async_pool.futures_per_thread",
+impl Clone for ThreadMetrics {
+    fn clone(&self) -> Self {
+        Self {
+            polled_futures: AtomicU64::new(self.polled_futures.load(Ordering::SeqCst)),
         }
     }
 }
+
+/// Inner state of asynchronous pool metrics.
+///
+/// This struct is not intended for direct use by end users. It is wrapped by [`AsyncPoolMetrics`]
+/// to provide a safe, ergonomic interface for tracking pool performance.
+#[derive(Debug)]
+pub struct Inner {
+    /// The name of the pool from which the metrics originate.
+    pool_name: &'static str,
```

Comment:
Decided to have it here, to avoid tracking it separately. If we were to need other tags in the future, we could generalize this into a `Tags` struct which is returned when emitting a metric.